external-crates/move/move-compiler/src/hlir/detect_dead_code.rs

```diff
@@ -339,120 +380,86 @@ fn tail_block(context: &mut Context, seq: &T::Sequence) -> Option<ControlFlow> {
+                    if cflow.value == aflow.value {
+                        context.report_value_error(sp(*eloc, cflow.value));
+                    } else {
+                        context.report_value_error(cflow);
+                        context.report_value_error(aflow);
+                    }
+                    return already_reported(*eloc);
                 }
-                _ => None,
             }
+            None
         }
         E::While(_, _) | E::Loop { .. } => statement(context, e),
         E::Block(seq) => value_block(context, seq),
 
         // -----------------------------------------------------------------------------------------
         //  calls and nested expressions
         // -----------------------------------------------------------------------------------------
-        E::ModuleCall(call) => {
-            let arg_flows = value_list(context, &call.arguments);
-            let result = if arg_flows.is_empty() {
-                None
-            } else {
-                already_reported(*eloc)
-            };
-            for flow in arg_flows {
-                context.report_value_error(flow);
-            }
-            result
-        }
-
-        E::Builtin(_, args) | E::Vector(_, _, _, args) => {
-            let arg_flows = value_list(context, args);
-            let result = if arg_flows.is_empty() {
-                None
-            } else {
-                already_reported(*eloc)
-            };
-            for flow in arg_flows {
-                context.report_value_error(flow);
-            }
-            result
-        }
-
-        E::Pack(_, _, _, fields) => {
-            let mut divergent_term = false;
-            for (_, _, (_, (_, field_exp))) in fields {
-                if let Some(control_flow) = value(context, field_exp) {
-                    context.report_value_error(control_flow);
-                    divergent_term = true;
-                }
-            }
-            if divergent_term {
-                already_reported(*eloc)
-            } else {
-                None
-            }
-        }
-
-        E::ExpList(items) => {
-            let mut divergent_term = false;
-            for item in items {
-                match item {
-                    T::ExpListItem::Single(entry, _) => {
-                        if let Some(control_flow) = value(context, entry) {
-                            context.report_value_error(control_flow);
-                            divergent_term = true;
+        E::ModuleCall(call) => value_report!(&call.arguments),
+
+        E::Builtin(_, args) | E::Vector(_, _, _, args) => value_report!(args),
+
+        E::Pack(_, _, _, fields) => fields
+            .iter()
+            .find_map(|(_, _, (_, (_, field_exp)))| value_report!(field_exp)),
+
+        E::ExpList(_) => {
+            use T::UnannotatedExp_ as TE;
+            if let TE::ExpList(ref items) = e.exp.value {
```

Comment:
minor nit, I tend to avoid `ref` in patterns because I think it is icky lol
No real opinions here though
```suggestion
            if let TE::ExpList(items) = &e.exp.value {
```